// mongodb/lib/operations/search_indexes/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateSearchIndexOperation = void 0;
const command_1 = require("../command");
const operation_1 = require("../operation");
/** @internal */
class UpdateSearchIndexOperation extends command_1.CommandOperation {
    constructor(collection, name, definition, options = {}) {
        super(collection, options);
        this.collection = collection;
        this.name = name;
        this.definition = definition;
        this.ns = collection.fullNamespace;
    }
    get commandName() {
        return "updateSearchIndex";
    }
    buildCommand(_connection, _session) {
        const namespace = this.collection.fullNamespace;
        return {
            updateSearchIndex: namespace.collection,
            name: this.name,
            definition: this.definition,
        };
    }
    buildOptions(timeoutContext) {
        return { session: this.session, timeoutContext };
    }
}
exports.UpdateSearchIndexOperation = UpdateSearchIndexOperation;
(0, operation_1.defineAspects)(UpdateSearchIndexOperation, [
    operation_1.Aspect.WRITE_OPERATION,
    operation_1.Aspect.RETRYABLE,
]);
